File: LeetCode/[M] - 438. Find All Anagrams in a String/solution.py
from typing import List


class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        if len(s) < len(p):
            return []
        wind, lay = {}, {}
        for i in range(len(p)):
            wind[s[i]] = wind.get(s[i], 0) + 1
            lay[p[i]] = lay.get(p[i], 0) + 1
        res = []
        if wind == lay:
            res.append(0)
        l = 0
        for r in range(len(p), len(s)):
            wind[s[r]] = wind.get(s[r], 0) + 1
            wind[s[l]] -= 1
            if wind[s[l]] == 0:
                del wind[s[l]]
            l += 1
            if wind == lay:
                res.append(l)
        return res

File: LeetCode/[M] - 438. Find All Anagrams in a String/solution.go
package main

import "maps"

func findAnagrams(s string, p string) []int {
	if len(s) < len(p) {
		return nil
	}
	wind, lay := make(map[byte]int, len(p)), make(map[byte]int, len(p))
	for i := range len(p) {
		wind[s[i]]++
		lay[p[i]]++
	}
	var res []int
	if maps.Equal(wind, lay) {
		res = append(res, 0)
	}
	l := 0
	for r := len(p); r < len(s); r++ {
		wind[s[r]]++
		wind[s[l]]--
		if wind[s[l]] == 0 {
			delete(wind, s[l])
		}
		l++
		if maps.Equal(wind, lay) {
			res = append(res, l)
		}
	}
	return res
}
